warehouse/csp.py

```diff
@@ -0,0 +1,58 @@
+import collections
+import copy
+
+
+def _serialize(policy):
+    return "; ".join([
+        " ".join([k] + [v2 for v2 in v if v2 is not None])
+        for k, v in sorted(policy.items())
+        if [v2 for v2 in v if v2 is not None]
+    ])
+
+
+def content_security_policy_tween_factory(handler, registry):
+    def content_security_policy_tween(request):
+        resp = handler(request)
+
+        try:
+            policy = request.find_service(name="csp")
+        except ValueError:
+            policy = collections.defaultdict(list)
+
+        # We don't want to apply our Content Security Policy to the debug
+        # toolbar, that's not part of our application and it doesn't work with
+        # our restrictive CSP.
+        if not request.path.startswith("/_debug_toolbar/") and policy:
+            resp.headers["Content-Security-Policy"] = _serialize(policy)
```

Comment:
This should be ``_serialize(policy).format(request=request)``. We use a slightly dynamic CSP in development based on the request attributes. So right now this CSP is:

```
$ curl -I http://192.168.99.100/
HTTP/1.1 200 OK
Server: gunicorn/19.4.5
Date: Sat, 05 Mar 2016 16:10:54 GMT
Connection: keep-alive
Content-Type: text/html; charset=UTF-8
ETag: "zzqeSeiFKitpuq9Sj56MPw"
Content-Length: 11602
Content-Security-Policy: connect-src 'self'; default-src 'none'; font-src 'self' fonts.gstatic.com; frame-ancestors 'none'; img-src 'self' {request.scheme}://{request.domain}:9000/ https://secure.gravatar.com; referrer origin-when-cross-origin; reflected-xss block; script-src 'self'; style-src 'self' fonts.googleapis.com
Vary: Accept-Encoding
```

Which we obviously don't want values like ``{request.scheme}://{request.domain}:9000/``.